docs: Add E2E testing tutorial. (#2591)
decred/politeiagui: docs/e2e.md

changes are applied on the list page, the tests should automatically detect them
if they mess up with the user expected app response.

With this in mind, a good test is the one that handles the expected application
UX for what the user is interacting with, including both successful and failed
actions.

Let's write a simple test case for our proposals list page:

```javascript
describe("Proposals List", () => {
  it("should render proposals list", () => {
    cy.visit("/");
    cy.findByTestId("proposal-title").should("be.visible");
    cy.findByTestId("proposal-date").should("be.visible");
  });
});
```

The `describe` method is used for grouping, and `it` is used for the
test case. Each test case is followed by assertions, which are used to control
the expected user behavior. As you can see, we are making 2
[assertions](https://docs.cypress.io/guides/core-concepts/introduction-to-cypress#Assertions)
on our test case using the
[`should`](https://docs.cypress.io/api/commands/should) command, which accepts
[Chai chainers](https://docs.cypress.io/guides/references/assertions#Chai) as an
argument.

> Check the Cypress and Jest docs for more info.

> However, the most applications are not static, and you can't test it only

Reviewer: ```suggestion
However, most of applications are not static, and you can't test them only
```

       // ...
     });
     // ...
   });
   ```

4. Now you can test pages that call the `/ticketvote/v1/inventory` without being
   connected to the ticketvote api.

---

Let's build our proposals list page test example using the created ticketvote
inventory middleware. First, let's keep in mind that the proposals page also
requests the `api/records/v1/records` endpoint, so let's assume we have a
`records.records` middleware, which will return a random records batch for
given inventory tokens.

```javascript
describe("Proposals list", () => {
  it("should render proposals list", () => {
    // middlewares
    cy.middleware("ticketvote.inventory");
    cy.middleware("records.records");
    // user behavior
    cy.visit(`/`);
    cy.wait("@ticketvote.inventory");
    cy.wait("@records.records");
    // assert that the proposal is fetched on the list
    cy.findAllByTestId("record-title").should("have.length", 1);
    // Scroll to bottom to render more proposals
    cy.scrollTo("bottom");
    cy.wait(1000);
    // make sure no new proposals were fetched, since the inventory has only
    // one token
    cy.findAllByTestId("record-title").should("have.length", 1);
  });
});
```

> Now, if we turned off the backend, the tests would still be able to render the

Reviewer 2: I think you could use the present tense here.
